command/agent/config/config.go

```diff
@@ -169,14 +171,253 @@ func NewConfig() *Config {
+		result.Vault = c2.Vault
+	}
+
+	result.PidFile = c.PidFile
+	if c2.PidFile != "" {
+		result.PidFile = c2.PidFile
+	}
+
+	return result
+}
+
+// ValidateConfig validates an Agent configuration after it has been fully merged together, to
+// ensure that required combinations of configs are there
+func (c *Config) ValidateConfig() error {
+	if c.APIProxy != nil && c.Cache != nil {
+		if c.Cache.UseAutoAuthTokenRaw != nil {
+			if c.APIProxy.UseAutoAuthTokenRaw != nil {
+				return fmt.Errorf("use_auto_auth_token defined in both api_proxy and cache config. Please remove this configuration from the cache block")
+			} else {
+				c.APIProxy.ForceAutoAuthToken = c.Cache.ForceAutoAuthToken
+			}
+		}
+	}
+
+	if c.Cache != nil {
+		if len(c.Listeners) < 1 && len(c.Templates) < 1 {
+			return fmt.Errorf("enabling the cache requires at least 1 template or 1 listener to be defined")
+		}
+
+		if c.Cache.UseAutoAuthToken {
+			if c.AutoAuth == nil {
+				return fmt.Errorf("cache.use_auto_auth_token is true but auto_auth not configured")
+			}
+			if c.AutoAuth != nil && c.AutoAuth.Method != nil && c.AutoAuth.Method.WrapTTL > 0 {
+				return fmt.Errorf("cache.use_auto_auth_token is true and auto_auth uses wrapping")
+			}
+		}
+	}
+
+	if c.APIProxy != nil {
+		if len(c.Listeners) < 1 {
+			return fmt.Errorf("configuring the api_proxy requires at least 1 listener to be defined")
+		}
+
+		if c.APIProxy.UseAutoAuthToken {
+			if c.AutoAuth == nil {
+				return fmt.Errorf("api_proxy.use_auto_auth_token is true but auto_auth not configured")
+			}
+			if c.AutoAuth != nil && c.AutoAuth.Method != nil && c.AutoAuth.Method.WrapTTL > 0 {
+				return fmt.Errorf("api_proxy.use_auto_auth_token is true and auto_auth uses wrapping")
+			}
+		}
+	}
+
+	if c.AutoAuth != nil {
+		if len(c.AutoAuth.Sinks) == 0 &&
+			(c.APIProxy == nil || !c.APIProxy.UseAutoAuthToken) &&
+			len(c.Templates) == 0 {
+			return fmt.Errorf("auto_auth requires at least one sink or at least one template or api_proxy.use_auto_auth_token=true")
+		}
+	}
+
+	if c.AutoAuth == nil && c.Cache == nil && len(c.Listeners) == 0 {
+		return fmt.Errorf("no auto_auth, cache, or listener block found in config")
+	}
+
+	return nil
+}
+
 // LoadConfig loads the configuration at the given path, regardless if
-// its a file or directory.
+// it's a file or directory.
 func LoadConfig(path string) (*Config, error) {
 	fi, err := os.Stat(path)
 	if err != nil {
 		return nil, err
 	}
 
+	if fi.IsDir() {
+		return LoadConfigDir(path)
+	}
+	return LoadConfigFile(path)
+}
+
+// LoadConfigDir loads the configuration at the given path if it's a directory
+func LoadConfigDir(dir string) (*Config, error) {
+	f, err := os.Open(dir)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	fi, err := f.Stat()
+	if err != nil {
+		return nil, err
+	}
+	if !fi.IsDir() {
+		return nil, fmt.Errorf("configuration path must be a directory: %q", dir)
+	}
+
+	var files []string
+	err = nil
+	for err != io.EOF {
```

Comment:
It's fine as you have it, though I think it might be a little more readable if you used filepath.WalkDir instead of Readdir here.